src/realm/object-store/audit.mm

```diff
@@ -1081,45 +1093,83 @@ void validate_metadata(std::vector<std::pair<std::string, std::string>>& metadat
 
 {
     util::CheckedLockGuard lock(m_mutex);
-    auto scope = std::make_shared<Scope>(Scope{m_metadata, std::string(activity)});
+    // scope id isn't used for this scope, so it's just an arbitrary value
+    auto scope = std::make_shared<Scope>(Scope{0, m_metadata, std::string(activity)});
     scope->events.push_back(audit_event::Custom{now(), std::string(activity), event_type, data});
     scope->completion = std::move(completion);
     trigger_write(std::move(scope));
 }
 
 void AuditContext::pin_version(VersionID version)
 {
-    for (auto& transaction : m_current_scope->source_transactions) {
-        if (transaction->get_version() == version.version)
-            return;
+    TransactionRef pin;
+    for (auto& scope : m_current_scopes) {
+        bool has_version =
+            std::any_of(scope->source_transactions.begin(), scope->source_transactions.end(), [&](auto& tr) {
+                return tr->get_version() == version.version;
+            });
+        if (!has_version) {
+            if (!pin) {
+                pin = m_source_db->start_read(version);
+            }
+            scope->source_transactions.push_back(pin);
+        }
+    }
+}
+
+uint64_t AuditContext::begin_scope(std::string_view name)
+{
+    util::CheckedLockGuard lock(m_mutex);
+    auto id = ++m_scope_counter;
+    m_logger->trace("Events: Beginning event scope '%1' on '%2' named '%3'", id, m_source_db->get_path(), name);
+    m_current_scopes.push_back(std::make_shared<Scope>(Scope{id, m_metadata, std::string(name)}));
```

Comment:
Since begin_scope now don't enforce the check, how you'd tell if user code simply didn't forget to end some of previous scopes? Also, are you sure that the reported situation in user code is not this exact issue really? With this design m_current_scopes may simply grow indefinitely. To me it seems that this previous API wasn't designed for such nested and overlapping usecases, so i don't think it's the way to go. Swift API for events also matches this methods one to one essentially, so the same problem will be introduced there.

Reply:
[The SDK API](https://github.com/realm/realm-swift/pull/8123) is changing to return a Scope which cleans up on deallocation. Propagating that down to the core level would merely make the SDK implementation more complicated, as we can't just store a std::shared_ptr on the Scope struct.